### PR TITLE
TOF DCS Processor

### PR DESCRIPTION
**Pull request by chiarazampolli**
This is the first implementation of the TOF DCS Processor. 
Still to-do:
- read DPIDs needed by TOF from CCDB
- fill CCDB object
- process the FEACSTATUS

### CODE REVIEW

**Comment by chiarazampolli** [comment]
Hello @aphecetche ,

While implementing the part for TOF, I made a couple of changes to the Generator. 
- L30 (old), 32 (new) of  Detectors/DCS/src/DataPointGenerator.cxx : this is a fix. Without it, the time was random, as "seconds" was uninitialized;
- L35 (old), 37 (new) of  Detectors/DCS/src/DataPointGenerator.cxx : this is to be able to pass a TDatime, which I need to change the timestamp of the generated DPs (otherwise it was always the same), see  Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h 
- From L71 (old) 72 (new) of  Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h : I use the TFID to change the timestamp of the generated points. In principle, each point will come with a different timestamp, so also this is a simplification. But what matters is that different maps from DCS have different timestamps for the same DPs
- in  Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h , I added TOF (few DPs, just for testing)

Please check that it is fine with you.

Chiara

File: Detectors/DCS/src/DataPointGenerator.cxx [hunk 1]
```diff
@@ -23,16 +24,28 @@ namespace
 {
 std::pair<uint32_t, uint16_t> getDate(const std::string& refDate)
 {
+
   uint32_t seconds;
   if (refDate.empty()) {
     auto current = std::time(nullptr);
     auto t = std::localtime(&current);
-    uint32_t seconds = mktime(t);
+    seconds = mktime(t);
```
Comment (chiarazampolli) [inline Detectors/DCS/src/DataPointGenerator.cxx:32]: @aphecetche : this is the fix.

File: Detectors/DCS/src/DataPointGenerator.cxx [hunk 2]
```diff
@@ -23,16 +24,28 @@ namespace
...
   } else {
     std::tm t{};
     std::istringstream ss(refDate);
     ss >> std::get_time(&t, "%Y-%b-%d %H:%M:%S");
-    seconds = mktime(&t);
+    if (ss.fail()) { // let's see if it was passed as a TDatime
```
Comment (chiarazampolli) [inline Detectors/DCS/src/DataPointGenerator.cxx:37]: @aphecetche : this is to accept also the TDatime format

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 3]
```diff
@@ -68,12 +69,20 @@ std::vector<int> generateIntegers(size_t size, int min, int max)
...
 std::vector<o2::dcs::DataPointCompositeObject> generate(const std::vector<HintType> hints,
-                                                        float fraction = 1.0)
+                                                        float fraction = 1.0,
+							uint64_t tfid = 0)
 {
   std::vector<o2::dcs::DataPointCompositeObject> dataPoints;
 
-  auto GenerateVisitor = [](const auto& t) {
-    return o2::dcs::generateRandomDataPoints({t.aliasPattern}, t.minValue, t.maxValue);
+  TDatime d;
```
Comment (chiarazampolli) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:87]: @aphecetche : this is to be able to change the timestamp at every map that we send, using the TF id.

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 4]
```diff
@@ -135,12 +156,14 @@ class DCSRandomDataGenerator : public o2::framework::Task
     // fraction is one if we generate FBI (Full Buffer Image)
     float fraction = (generateFBI ? 1.0 : mDeltaFraction);
 
-    auto dpcoms = generate(mDataPointHints, fraction);
+    TDatime d; 
+    auto dpcoms = generate(mDataPointHints, fraction, tfid);
+    auto tofdpcoms = generate(mTOFDataPointHints, fraction, tfid);
 
-    // the output must always get both FBI and Delta, but one of them is empty.
+    LOG(INFO) << "***************** TF " << tfid << " has generated " << tofdpcoms.size() << " DPs for TOF";
```
Comment (chiarazampolli) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:180]: @shahor02 , @aphecetche : DCS will send only 1 map at a time, either FBI or delta.

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 5]
```diff
@@ -56,8 +56,14 @@ using HintType = std::variant<DataPointHint<double>,
 {
   std::uniform_int_distribution<int> distribution(min, max);
   std::mt19937 generator;
-  std::vector<int> data(size);
-  std::generate(data.begin(), data.end(), [&]() { return distribution(generator); });
+  std::vector<int> data;
```
Comment (chiarazampolli) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:60]: @aphecetche , this is to make the generation of the DPs in the delta map such that there are no duplicated DPs - the delta map will contain one value only per DP.

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 6]
```diff
@@ -55,7 +55,7 @@ using HintType = std::variant<DataPointHint<double>,
   std::vector<int> generateIntegers(size_t size, int min, int max)
 {
   std::uniform_int_distribution<int> distribution(min, max);
-  std::mt19937 generator;
+  std::mt19937 generator(std::random_device{}());
```
Comment (chiarazampolli) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:59]: @aphecetche : this is to change the subset of DPs that are generated at every TF

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 7]
```diff
@@ -135,12 +168,14 @@ class DCSRandomDataGenerator : public o2::framework::Task
...
     float fraction = (generateFBI ? 1.0 : mDeltaFraction);
 
-    auto dpcoms = generate(mDataPointHints, fraction);
+    TDatime d; 
+    auto dpcoms = generate(mDataPointHints, fraction, tfid);
+    auto tofdpcoms = generate(mTOFDataPointHints, fraction, tfid);
 
-    // the output must always get both FBI and Delta, but one of them is empty.
+    LOG(INFO) << "***************** TF " << tfid << " has generated " << tofdpcoms.size() << " DPs for TOF";
     std::vector<o2::dcs::DataPointCompositeObject> empty;
```
Comment (shahor02) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h]: this seems to be never used.

**Comment by chiarazampolli** [comment]
Dear @noferini , @shahor02 , @aphecetche ,

I am done with the changes I wanted to make to this PR, so I removed the WIP from the title. As soon as the basic checks will pass, can you review it?

Thanks!

Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 , @noferini , @aphecetche ,

The only error I could find in https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/5047/e6413573c142517866c4bca5f7f2174b881f7beb/build_O2_o2/fullLog.txt is:

> Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libRIO, output: $ROOT_DYN_PATH/libRIO
Error in <TSystem::ExpandFileName>: input: $ROOT_DYN_PATH/libCling, output: $ROOT_DYN_PATH/libCling

But it is not related to this PR. Any idea? Is there something else?

Chiara

**Comment by sawenzel** [comment]
@chiarazampolli : This error is annoying but not critical. It just means that the variable ROOT_DYN_PATH (announced to ROOT somewhere) is not set in the compilation environment.
As far as I can tell, the build is 100% ok. Maybe @TimoWilken can comment why one of them is flagged as failed.

**Comment by chiarazampolli** [comment]
Ciao @sawenzel ,
Thanks! I did not spot any other error, is it normal that this one triggers the failure of o2 build?
Chiara

**Comment by shahor02** [comment]
Hi Chiara,

It reports a failure due to the "Manually-specified variables were not used by the project:" warning in cmake, see the entry page of https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/5047/e6413573c142517866c4bca5f7f2174b881f7beb/build_O2_o2/pretty.html.
Just ignore it.

Cheers
  Ruben

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by chiarazampolli** [comment]
Thanks Ruben!

Chiara

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h [hunk 8]
```diff
@@ -68,12 +79,20 @@ std::vector<int> generateIntegers(size_t size, int min, int max)
...
-    return o2::dcs::generateRandomDataPoints({t.aliasPattern}, t.minValue, t.maxValue);
+  TDatime d;
+  auto dsec = d.Convert();
+  dsec += tfid;
+  d.Set(dsec);
+
+  std::string refDate = d.AsString();
+
+  auto GenerateVisitor = [refDate](const auto& t) {
+    return o2::dcs::generateRandomDataPoints({t.aliasPattern}, t.minValue, t.maxValue, refDate);
```
Comment (shahor02) [inline Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:95]: L106-107 makes unnecessary copy, instead of 
````
auto tmp = dataPoints;
dataPoints.clear();`
````
 better do

````
std::vector<o2::dcs::DataPointCompositeObject> tmp;
tmp.swap(dataPoints);
````

File: Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+// or submit itself to any jurisdiction.
+
+#include "DetectorsDCS/DataPointIdentifier.h"
+#include "DetectorsDCS/DataPointValue.h"
+#include "Framework/TypeTraits.h"
+#include <unordered_map>
+namespace o2::framework
+{
+template <>
+struct has_root_dictionary<std::unordered_map<o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue>, void> : std::true_type {
```
Comment (shahor02) [inline Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx]: why is this needed here? Does not ``class std::unordered_map < o2::dcs::DataPointIdentifier, o2::dcs::DataPointValue> + ;`` do the job?

File: Detectors/TOF/calibration/include/TOFCalibration/TOFDCSProcessor.h [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    lastValue = std::make_pair(0, -999999999);
+    midValue = std::make_pair(0, -999999999);
+    maxChange = std::make_pair(0, -999999999);
+  }
+  void makeEmpty()
+  {
+    firstValue.first = lastValue.first = midValue.first = maxChange.first = 0;
+    firstValue.second = lastValue.second = midValue.second = maxChange.second = -999999999;
+  }
+  void print() const
```
Comment (shahor02) [inline Detectors/TOF/calibration/include/TOFCalibration/TOFDCSProcessor.h]: Unless strictly needed, IO methods better keep in cxx file (together with Logger header).

File: Detectors/TOF/calibration/macros/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+# verbatim in the file "COPYING".
+#
+# See http://alice-o2.web.cern.ch/license for full licensing information.
+#
+# In applying this license CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+
+if(BUILD_TESTING)
+  o2_add_test(
```
Comment (shahor02) [inline Detectors/TOF/calibration/macros/CMakeLists.txt]: Just to be sure what you want to do here: you declare this to be a test, but the ``makeCCDBEntryForDCS.cxx`` is a simple executable, w/o any BOOST test, so it will never fail.

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+int TOFDCSProcessor::process(const gsl::span<const DPCOM> dps)
+{
+
+  // first we check which DPs are missing - if some are, it means that
+  // the delta map was sent
+  if (mVerbose) {
+    LOG(INFO) << "\n\n\nProcessing new TF\n-----------------";
+  }
+  if (!mStartTFset) {
+    mStartTF = mTF;
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx:67]: add allso ``mStartTFset = true;`` otherwise it will always update ``mStartTF``

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+    mapin[it.id] = it.data;
+  }
+  for (auto& it : mPids) {
+    const auto& el = mapin.find(it.first);
+    if (el == mapin.end()) {
+      LOG(DEBUG) << "DP " << it.first << " not found in map";
+    } else {
+      LOG(DEBUG) << "DP " << it.first << " found in map";
+    }
+  }
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx:82]: why is this needed?

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+    } else if (type == RAW_INT) {
+      LOG(INFO);
+      LOG(INFO) << "Processing DP = " << dpcom << ", with value = " << o2::dcs::getValue<int32_t>(dpcom);
+    }
+  }
+  auto flags = val.get_flags();
+  if (processFlags(flags, dpid.get_alias()) == 0) {
+    // now I need to access the correct element
+    if (type == RAW_DOUBLE) {
+      // for these DPs, we will store the first, last, mid value, plus the value where the maximum variation occurred
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx:133]: each map query ``mDpsdoublesmap[dpid]`` takes some cpu time, better do this once as
````
auto& dvec = mDpsdoublesmap[dpid];
````
and use it below.

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+        mDpsdoublesmap[dpid].push_back(val);
+      }
+    }
+
+    if (type == RAW_INT) {
+      // for these DPs, we need some processing
+      if (std::strstr(dpid.get_alias(), "FEACSTATUS") != nullptr) { // DP is FEACSTATUS
+        std::string aliasStr(dpid.get_alias());
+        // extracting DDL number (regex is quite slow, using this way)
+        std::size_t const nn = aliasStr.find_first_of("0123456789");
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: you can make it even faster by doing
````
const offs = sizeof("TOF_FEACSTATUS");
std::size_t const nn = aliasStr.find_first_of("0123456789", offs);
````

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+        if (mVerbose) {
+          LOG(INFO) << "Updating previous FEAC status for DDL " << iddl;
+        }
+        mPrevFEACstatus[iddl] = feacstatus;
+      } // end processing current DP, when it is of type FEACSTATUS
+
+      if (std::strstr(dpid.get_alias(), "HVSTATUS") != nullptr) { // DP is HVSTATUS
+        std::string aliasStr(dpid.get_alias());
+        // extracting SECTOR and PLATE number (regex is quite slow, using this way)
+        std::size_t const nn = aliasStr.find_first_of("0123456789");
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: idem, faster to provide an offset for ``find_first_of``

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx [hunk 17]
```diff
@@ -0,0 +1,555 @@
...
+  union Converter {
+    uint64_t raw_data;
+    double double_value;
+  } converter0, converter1;
+
+  for (const auto& it : mPids) {
+    if (it.second == true) { // we processed the DP at least 1x
+      const auto& type = it.first.get_type();
+      if (type == o2::dcs::RAW_DOUBLE) {
+        mTOFDCS[it.first].firstValue.first = mDpsdoublesmap[it.first][0].get_epoch_time();
```
Comment (shahor02) [inline Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: here too, better get it once as ``auto& tofcds = mTOFDCS[it.first];``, then use it below.

File: Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+    sendOutput(pc.outputs());
+  }
+
+  void endOfStream(o2::framework::EndOfStreamContext& ec) final
+  {
+    mProcessor->finalize();
+    sendOutput(ec.outputs());
+    const auto& payload = mProcessor->getTOFDPsInfo();
+    auto& info = mProcessor->getccdbDPsInfo();
+    auto image = o2::ccdb::CcdbApi::createObjectImage(&payload, &info);
```
Comment (shahor02) [inline Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h:113]: so, the double DPs will be saved only once in the very end of the workflow lifetime, which might be large. Is this what you want?

**Review by shahor02 — CHANGES_REQUESTED** [review]
Hi @chiarazampolli 

Please see a few comments below (only 1 is a real bug).

File: Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h (hunk 8 above)
Comment (chiarazampolli) [reply Detectors/DCS/testWorkflow/DCSRandomDataGeneratorSpec.h:95]: Done.

File: Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx (hunk 9 above)
Comment (chiarazampolli) [reply Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx]: If I don't remember wrong, we had issues with unordered_map, to pass it with DPL. Since now we pass a vector, maybe it is not needed. I will try.

File: Detectors/TOF/calibration/include/TOFCalibration/TOFDCSProcessor.h (hunk 10 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/include/TOFCalibration/TOFDCSProcessor.h]: Done

File: Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx (hunk 9 above)
Comment (chiarazampolli) [reply Detectors/DCS/testWorkflow/dcs-sim-workflow.cxx]: Removed

File: Detectors/TOF/calibration/macros/CMakeLists.txt (hunk 11 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/macros/CMakeLists.txt]: I actually did not want to test this, but it was complaining if I did not. 
Should I simply add it to the exceptions?

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 12 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx:67]: Yes! I forgot, thanks for noticing!

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 14 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx:133]: done

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 15 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: Actually, I should do either:
```
	const auto offs = std::strlen("TOF_FEACSTATUS");
```
or 
``` 
const auto offs = sizeof(""TOF_FEACSTATUS") - 1;
```
because it is null terminated. I used strlen, is it ok?

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 16 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: see above

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 17 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: done, plus bug fix (I should process in the finalize only the DPs of type DOUBLE), plus improvements.

File: Detectors/TOF/calibration/macros/CMakeLists.txt (hunk 11 above)
Comment (shahor02) [reply Detectors/TOF/calibration/macros/CMakeLists.txt]: if you wanted to simply have an executable, then use ``o2_add_executable`` instead of ``o2_add_test``. Also, you probably don't need ``if(BUILD_TESTING)``

File: Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h (hunk 18 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h:113]: For now, I think it is ok, but we can change it to have it ever n TFs. I will do this modification in another PR, if you agree.

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 15 above)
Comment (shahor02) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: ok, though you have an "_" after ``TOF_FEACSTATUS``, so like that you will start from it.

File: Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h (hunk 18 above)
Comment (shahor02) [reply Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h:113]: sure

File: Detectors/TOF/calibration/macros/CMakeLists.txt (hunk 11 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/macros/CMakeLists.txt]: done!

File: Detectors/TOF/calibration/src/TOFDCSProcessor.cxx (hunk 15 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/src/TOFDCSProcessor.cxx]: right, my typo. fixed.

File: Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h (hunk 18 above)
Comment (chiarazampolli) [reply Detectors/TOF/calibration/testWorkflow/TOFDCSDataProcessorSpec.h:113]: Ok, will take care in January :-)

**Comment by chiarazampolli** [comment]
A side remark about something weird: after running clang-format, I still get an error, in a file that I think clang-format itself changed at the previous iteration.

**Comment by TimoWilken** [comment]
Hi @chiarazampolli! I'm trying to track down some other weirdness in the clang-format checker. Do you happen to have a log with that error somewhere? It might be useful...

**Comment by chiarazampolli** [comment]
Ciao @TimoWilken , 

Unfortunately not, the logs were those from the CI, so they are lost now. 

Chiara

**Comment by TimoWilken** [comment]
Hi @chiarazampolli, fair enough! Please let me know if you see it happen again.

**Comment by chiarazampolli** [comment]
Ciao @TimoWilken , 

Sure!

Chiara

**Review by shahor02 — APPROVED** [review]

**Comment by shahor02** [comment]
@chiarazampolli  @noferini shall I merge it or Francesco wants to review it?

**Comment by chiarazampolli** [comment]
Hello,

@noferini  is away, I think, until end of the holidays. Maybe we could merge it, and in case make modifications? In this way we can send it around as an example (and the README also can be useful).

Chiara
